### ISSUE TITLE
Serverless complains about "Rest API id could not be resolved."

### ISSUE THREAD

**Issue by gevorggalstyan** [serverless/serverless#6607]
# This is a Bug Report

## Description

- What went wrong?
After updating serverless to version 1.51.0 the deployment process never succeeds. Deploy command does the job partially. I can see in the logs that Stack update has finished and it prints out the endpoints and the functions, but then immediately fires an error that looks like this:

```txt
Serverless Error ---------------------------------------
 
  Rest API id could not be resolved.
  This might be caused by a custom API Gateway configuration.
  
  In given setup stage specific options such as `tracing`, `logs` and `tags` are not supported.
  
  Please update your configuration (or open up an issue if you feel that there's a way to support your setup).
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              10.16.0
     Framework Version:         1.51.0
     Plugin Version:            1.3.9
     SDK Version:               2.1.0
```

- What did you expect should have happened?
The deployment should pass without any issues like it was with the previous version (1.50.1)

- What was the config you used?
     Operating System:          darwin
     Node Version:              10.16.0
     Framework Version:         1.51.0
     Plugin Version:            1.3.9
     SDK Version:               2.1.0

- What stacktrace or error message from your provider did you see?
```txt
Serverless Error ---------------------------------------
 
  ServerlessError: Rest API id could not be resolved.
  This might be caused by a custom API Gateway configuration.
  
  In given setup stage specific options such as `tracing`, `logs` and `tags` are not supported.
  
  Please update your configuration (or open up an issue if you feel that there's a way to support your setup).
      at resolveAccountId.call.then.then (/Users/g/Sites/app/node_modules/serverless/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js:66:17)
  From previous event:
      at AwsCompileApigEvents.updateStage (/Users/g/Sites/app/node_modules/serverless/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js:44:8)
      at Object.after:deploy:deploy [as hook] (/Users/g/Sites/app/node_modules/serverless/lib/plugins/aws/package/compile/events/apiGateway/index.js:85:28)
      at BbPromise.reduce (/Users/g/Sites/app/node_modules/serverless/lib/classes/PluginManager.js:464:55)
  From previous event:
      at PluginManager.invoke (/Users/g/Sites/app/node_modules/serverless/lib/classes/PluginManager.js:464:22)
      at PluginManager.run (/Users/g/Sites/app/node_modules/serverless/lib/classes/PluginManager.js:496:17)
      at variables.populateService.then (/Users/g/Sites/app/node_modules/serverless/lib/Serverless.js:116:33)
      at runCallback (timers.js:705:18)
      at tryOnImmediate (timers.js:676:5)
      at processImmediate (timers.js:658:5)
      at process.topLevelDomainCallback (domain.js:126:23)
  From previous event:
      at Serverless.run (/Users/g/Sites/app/node_modules/serverless/lib/Serverless.js:103:74)
      at serverless.init.then (/Users/g/Sites/app/node_modules/serverless/bin/serverless.js:60:28)
      at /Users/g/Sites/app/node_modules/graceful-fs/graceful-fs.js:136:16
      at /Users/g/Sites/app/node_modules/graceful-fs/graceful-fs.js:57:14
      at FSReqWrap.args [as oncomplete] (fs.js:140:20)
  From previous event:
      at initializeErrorReporter.then (/Users/g/Sites/app/node_modules/serverless/bin/serverless.js:60:6)
      at runCallback (timers.js:705:18)
      at tryOnImmediate (timers.js:676:5)
      at processImmediate (timers.js:658:5)
      at process.topLevelDomainCallback (domain.js:126:23)
  From previous event:
      at Object.<anonymous> (/Users/g/Sites/app/node_modules/serverless/bin/serverless.js:46:39)
      at Module._compile (internal/modules/cjs/loader.js:776:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
      at Module.load (internal/modules/cjs/loader.js:653:32)
      at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
      at Function.Module._load (internal/modules/cjs/loader.js:585:3)
      at Function.Module.runMain (internal/modules/cjs/loader.js:829:12)
      at findNodeScript.then.existing (/Users/g/.nodenv/versions/10.16.0/lib/node_modules/npm/node_modules/libnpx/index.js:268:14)
```

**Comment by medikoo** [comment]
@gevorggalstyan can you share your `serverless.yml` (at least part related to API Gateway configuration) ?

**Comment by exoego** [comment]
Sorry, accidentally  closed

**Comment by gevorggalstyan** [comment]
@medikoo I think I found the exact lines that cause this

```yaml
provider:
  logs:
    restApi: true
```
 We had these lines based on https://serverless.com/framework/docs/providers/aws/guide/serverless.yml/

**Comment by medikoo** [comment]
What do you have at `provider` and `apiGateway` config? Is there a chance you can share all config, with sensitive parts masked out?

**Comment by gevorggalstyan** [comment]
Does this help?
```yaml
frameworkVersion: ">=1.1.0 <2.0.0"
provider:
  name: aws
  runtime: nodejs10.x
  region: >-
    ${opt:region, self:custom.secrets.DEFAULT_REGION,
    self:provider.environment.DEFAULT_REGION}
  stage: "${opt:stage, self:custom.defaults.stage}"
  profile: "${self:custom.profiles.${self:provider.stage}}"
  timeout: 30 # The default is 6 seconds. Note: API Gateway current maximum is 30 seconds
  memorySize: 1024
  apiGateway:
    minimumCompressionSize: 1024
  tracing:
    apiGateway: true
    lambda: true
  logs:
    restApi: true

cors: &cors
  origin: "*"
  headers:
    - Content-Type
    - X-Amz-Date
    - Authorization
    - X-Api-Key
    - X-Amz-Security-Token
    - X-Amz-User-Agent
    # A custom header for provider portal to request cache clean if it is set to 0
    - X-Cache-Control
  allowCredentials: false
  maxAge: 86400
  # Caches on browser and proxy for 1 hour and doesn't allow proxy to serve out of date content
  cacheControl: "max-age=3600, s-maxage=3600, proxy-revalidate"

functions:
  koa:
    handler: sls/koa-sls-wrapper.ts
    memorySize: 2048
    events:
      # These would allow any request to be handled by koa
      # which we do not want to allow but keeping these here as a reference
      # - http: any /
      # - http: "any {proxy+}"

      # region Public API endpoints
      - http:
          cors: *cors
          method: any
          path: "migrations/{id}"
      # endregion
```

**Comment by medikoo** [comment]
@gevorggalstyan unfortunately with above example I wasn't able to reproduce the error. Here's a sample project I've prepared, and it deploys without issues: https://gist.github.com/medikoo/9a7091308770885b2314fc0c197cd9d9

This error appears when you rely on options which are not perfectly supported in CloudFormation, therefore we apply them via SDK calls (which happens after deployment of CloudFormation stack).

And for that calls we need to be able to resolve Rest API id programmatically. This resolution takes place here: https://github.com/serverless/serverless/blob/91ae8bcc17d3ab7874507c8192375b4a28627592/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js#L88-L96

Are you maybe configuring an additional custom `AWS::ApiGateway::Resource` in `resources.Resources` section of `serverless.yml`?

**Comment by gevorggalstyan** [comment]
@medikoo Our application has been deployed with previous versions successfully and even with version 1.50.1. The issue has only started with version 1.51.0. I can see in that version commits that there have been several changes related to logs at least. There is nothing else left in the configs, just a long list of -http events on the same function.

Can you please try deploying with version 1.50.1 and then redeploying with 1.51.0?

**Comment by gevorggalstyan** [comment]
Update: I have tried deploying the stack in your gist with 1.50.1 and then with 1.51.0 and did not hit the issue.

Going to redeploy the app from ground-up, to check if it works that way.

**Comment by gevorggalstyan** [comment]
@medikoo I found the exact commit on the master that is causing the issues for us:

https://github.com/serverless/serverless/commit/dda0c71048f90efe8d1c5b2c774dbbf6023b86b4

I think the problem is in the assumption that AWS::ApiGateway::Resource will/must always be in the main stack. In fact, it can be in pretty much any related nested stack. After the changes, the code tries to find it in the main stack's resources and fails.

Before this change, the code that was initializing the `apiName` was pretty much making a correct assumption.

```js
const apiName = provider.apiName || `${this.options.stage}-${this.state.service.service}`;
```

This is almost always correct if you do not define a custom rest API id.

After the change, this assumption has been removed entirely and the code looks like this

```js
if (!apiGatewayResource) {
      resolve(null);
      return;
}

const apiName = apiGatewayResource.Properties.Name;
```

So basically if we use any form of resource splitting and the API gateway resource happens to be in any stack other than the main one, the code will exit.

Do you think we could change the code to look something like this?

```js
    // Keeping the code for reference, it should be deleted
    // if (!apiGatewayResource) {
    //   resolve(null);
    //   return;
    // }
    const apiName = apiGatewayResource ? apiGatewayResource.Properties.Name : provider.apiName || `${this.options.stage}-${this.state.service.service}`;
```

I think this will be a good compromise and give more flexibility to the users of the framework.

**Comment by medikoo** [comment]
> I think the problem is in the assumption that AWS::ApiGateway::Resource will/must always be in the main stack. In fact, it can be in pretty much any related nested stack.

Indeed I didn't took into account that users may rely on nested stacks. Good find!

> Do you think we could change the code to look something like this?

Yes, definitely we can do that. Do you want to open a PR?

**Comment by gevorggalstyan** [comment]
Sure thing! Here it is https://github.com/serverless/serverless/pull/6611

Would be very thankful if we could get this merged as soon as possible so we can re-enable our logs. Thank you.
